activerecord/lib/active_record/associations/lazy_preloader.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  module Associations
+    class LazyPreloader
+      # Implements the details of lazy eager loading of Active Record associations.
+      #
+      # Suppose that you have the following two Active Record models:
+      #
+      #   class Author < ActiveRecord::Base
+      #     # columns: name, age
+      #     has_many :books
+      #   end
+      #
+      #   class Book < ActiveRecord::Base
+      #     # columns: title, sales, author_id
+      #   end
+      #
+      # When you load an author with all associated books Active Record will make
+      # only one query:
+      #
+      #   authors = Author.lazy_preload(:books).where(name: ['bell hooks', 'Homer']).to_a
+      #
+      #   => SELECT `authors`.* FROM `authors` WHERE `name` IN ('bell hooks', 'Homer')
+      #
+      # Then you probably will want to retrieve books while iterating authors
+      #
+      #   authors.each { |author| read author.books }
+      #
+      # And only after books method invocation second query will be executed
+      # to load all books for authors in the collection:
+      #
+      # => SELECT `books`.* FROM `books` WHERE `author_id` IN (2, 5)
+      #
+      class WeakRegistry
+        @@weakmap = ::ObjectSpace::WeakMap.new
```

Comment:
Not sure about `WeakMap`, it's considered to be private; Ruby docs suppose using [`WeakRef`](https://ruby-doc.org/stdlib-2.5.1/libdoc/weakref/rdoc/WeakRef.html) instead. But that would bring the complexity of cleaning up dead refs.

Maybe, a simpler alternative would be to store the preloader within the owner as an instance variable.

Reply:
Switched back to an instance variable.